auth/auth.go

```diff
@@ -0,0 +1,281 @@
+package auth
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-oidc"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"golang.org/x/oauth2"
+
+	"github.com/sourcegraph/sourcegraph-accounts-sdk-go/scopes"
+)
+
+// Config contains the configuration for the SAMS authentication handler.
+type Config struct {
+	// Issuer is the SAMS instance URL, e.g. "https://accounts.sourcegraph.com".
+	Issuer string
+	// ClientID is the SAMS client ID, e.g. "sams_cid_xxx".
+	ClientID string
+	// ClientSecret is the SAMS client secret, e.g. "sams_cs_xxx".
+	ClientSecret string
+	// RequestScopes is the list of requested scopes for access tokens that are
+	// issued to this client.
+	RequestScopes []scopes.Scope
+	// RedirectURI is the URL to redirect to after the user has authenticated.
```

Comment:
This has to match one of the RedirectURIs that was supplied to SAMS when registering the OAuth client, right? Probably worth calling that out. (i.e. that this cannot be set arbitrarily, it is essentially coupled to the client ID/secret.)

Reply:
Updated in https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/8/commits/9e9ced758726b7dbc703c326f59c93d41bc45b8c